### PR TITLE
feat: Implement hints on uint384 lib (Part 1)

### PR DESCRIPTION
**Pull request by fmoletta**
lib where these hints come from can be found here (https://github.com/NethermindEth/research-basic-Cairo-operations-big-integers/blob/main/lib/uint384.cairo)
Contains 5 hints:
UINT384_UNSIGNED_DIV_REM
UINT384_SPLIT_128
ADD_NO_UINT384_CHECK
UINT384_UNSIGNED_DIV_REM_EXPANDED
UINT384_SQRT

### CODE REVIEW

File: src/hint_processor/builtin_hint_processor/hint_code.rs [hunk 1]
```diff
@@ -617,5 +617,38 @@ from starkware.python.math_utils import recover_y
...
+pub(crate) const UINT384_UNSIGNED_DIV_REM_EXPANDED: &str = "def split(num: int, num_bits_shift: int, length: int):\n    a = []\n    for _ in range(length):\n        a.append( num & ((1 << num_bits_shift) - 1) )\n        num = num >> num_bits_shift\n    return tuple(a)\n\ndef pack(z, num_bits_shift: int) -> int:\n    limbs = (z.d0, z.d1, z.d2)\n    return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))\n\ndef pack2(z, num_bits_shift: int) -> int:\n    limbs = (z.b01, z.b23, z.b45)\n    return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))\n\na = pack(ids.a, num_bits_shift = 128)\ndiv = pack2(ids.div, num_bits_shift = 128)\nquotient, remainder = divmod(a, div)\n\nquotient_split = split(quotient, num_bits_shift=128, length=3)\nassert len(quotient_split) == 3\n\nids.quotient.d0 = quotient_split[0]\nids.quotient.d1 = quotient_split[1]\nids.quotient.d2 = quotient_split[2]\n\nremainder_split = split(remainder, num_bits_shift=128, length=3)\nids.remainder.d0 = remainder_split[0]\nids.remainder.d1 = remainder_split[1]\nids.remainder.d2 = remainder_split[2]";
+pub(crate) const UINT384_SQRT: &str = "from starkware.python.math_utils import isqrt
```
Comment (Oppen) [inline src/hint_processor/builtin_hint_processor/hint_code.rs:632]: Let's use the multiline string syntax so it keeps readable.

File: src/hint_processor/builtin_hint_processor/uint384.rs [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+ */
+pub fn uint384_sqrt(
+    vm: &mut VirtualMachine,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let a = pack(BigInt3::from_var_name("a", vm, ids_data, ap_tracking)?, 128);
+    let root_addr = get_relocatable_from_var_name("root", vm, ids_data, ap_tracking)?;
+    let root = isqrt(&a)?;
+    if root.is_zero() || root >= BigUint::one().shl(192_u32) {
```
Comment (Oppen) [inline src/hint_processor/builtin_hint_processor/uint384.rs:294]:
```suggestion
    if root.is_zero() || root.bits() > 192 {
```

**Review by Oppen — CHANGES_REQUESTED** [review]
Why are all the hint functions public?

**Comment by fmoletta** [comment]
> Why are all the hint functions public
No real reason, they have all been public since the start, we can create an issue to change all of them to pub(crate)

File: src/hint_processor/builtin_hint_processor/uint384.rs [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+/* Reduced version of Uint384_expand
+The full version has 7 limbs (B0, b01, b12, b23, b34, b45, b5), but only 3 are used by the pack2 fn (b01, b23, b45)
+As there are no other uses of Uint384_expand outside of these in the lib, we can use a reduced version with just 3 limbs
+*/
+#[derive(Debug, PartialEq)]
+#[allow(non_snake_case)]
+pub(crate) struct Uint384ExpandReduced<'a> {
+    pub b01: Cow<'a, Felt252>,
+    pub b23: Cow<'a, Felt252>,
+    pub b45: Cow<'a, Felt252>,
```
[the remark this answers is not in the document]
Comment (fmoletta) [reply src/hint_processor/builtin_hint_processor/uint384.rs:31]: Same problem as [here](https://github.com/lambdaclass/cairo-rs/pull/851#discussion_r1110282946), its not possible to use references here

**Review by Oppen — APPROVED** [review]

**Pull request merged**